charms/istio-pilot/src/charm.py

```diff
@@ -39,10 +39,13 @@ def __init__(self, *args):
         self.framework.observe(self.on.install, self.install)
         self.framework.observe(self.on.remove, self.remove)
 
+        self.framework.observe(self.on.config_changed, self.handle_default_gateways)
+
         self.framework.observe(self.on["istio-pilot"].relation_changed, self.send_info)
 
         self.framework.observe(self.on['ingress'].relation_changed, self.handle_ingress)
         self.framework.observe(self.on['ingress'].relation_departed, self.handle_ingress)
+        self.framework.observe(self.on['ingress'].relation_broken, self.handle_ingress)
```

Comment:
Should we be subscribed to both of these events?  Departed happens if a unit leaves a relation, and broken happens if all units leave the relation (and thus the relation is removed).  Do both have significance for us, or do we need only `relation_broken`?  I don't think we do anything different depending on the number of units on the other side

If both matter, should [this line](https://github.com/canonical/istio-operators/pull/43/files#diff-d3a4e684bd1a3a1822df5ab7605cfce00c97cd7b52f44ed770526c4f4f80eacdR148) catch `RelationDepartedEvent` as well?  I think no, as we don't want to delete the virtual service just because some (but not all) units of our related app have left, but thought I'd mention it just in case

Reply:
Departed only matters if per-unit routing is being used, but in that case it does matter, yes, since we want to remove that route from the VS (but not the entire VS).